### PR TITLE
[SNOW-1491284] DataFrame.{collect,collect_nowait,count} collectors

### PR DESCRIPTION
**Pull request by sfc-gh-lspiegelberg**
Support `DataFrame.{collect,collect_nowait,count}` for Snowpark IR. Furthermore, modifies test infrastructure to be able to pass multiple ASTs to the unparser in order so multiple evals can be unparsed. Adds new AstListener class for both server connection and mock server connection to capture ASTs easily. Modified steel-thread to work with AstListener to print out AST for interested audience.

### CODE REVIEW

File: tests/ast/data/DataFrame.collect.test [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+res1 = session.table("test_table")
+
+res2 = res1.collect()
+
+res2
+
+res3 = res1.collect(block=False)
+
+res3
+
```
Comment (sfc-gh-lspiegelberg) [inline tests/ast/data/DataFrame.collect.test:30]: need to fix here to use stringlit

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -638,12 +639,49 @@ def collect(
         See also:
             :meth:`collect_nowait()`
         """
+
+        if _emit_ast:
+            # Add an Assign node for SpDataframeCollect()
```
Comment (sfc-gh-azwiegincew) [inline src/snowflake/snowpark/dataframe.py:644]: Should be consistent with the comment in `show()`:
```
# Add an Assign node that applies SpDataframeShow() to the input, followed by its Eval.
```

File: src/snowflake/snowpark/dataframe.py [hunk 3]
```diff
@@ -666,13 +705,48 @@ def collect_nowait(
         See also:
             :meth:`collect()`
         """
+        if _emit_ast:
+            # Add an Assign node for SpDataframeCollect()
+            repr = self._session._ast_batch.assign()
+            expr = with_src_position(repr.expr.sp_dataframe_collect_no_wait)
```
Comment (sfc-gh-azwiegincew) [inline src/snowflake/snowpark/dataframe.py:711]: I suspect we should just add `no_wait: Bool` to `SpDataframeCollect` and factor out the code here.

File: src/snowflake/snowpark/dataframe.py [hunk 4]
```diff
@@ -666,13 +705,48 @@ def collect_nowait(
...
+
+        if self._session._conn.is_phase1_enabled():
+            # TODO: Logic here should be
+            # ast = self._session._ast_batch.flush()
+            # res = self._session._conn.ast_query(ast)
+            raise NotImplementedError(
+                "TODO: Implement collect() with EvalResult in Phase1."
+            )
+
+        # Phase0 flushes AST and encodes it as part of the query.
```
Comment (sfc-gh-azwiegincew) [inline src/snowflake/snowpark/dataframe.py:738]: Space between "Phase" and "0".

File: src/snowflake/snowpark/dataframe.py [hunk 5]
```diff
@@ -3418,13 +3498,48 @@ def count(
                 When it is ``False``, this function executes the underlying queries of the dataframe
                 asynchronously and returns an :class:`AsyncJob`.
         """
+
+        if _emit_ast:
+            # Add an Assign node for SpDataframeCollect()
```
Comment (sfc-gh-azwiegincew) [inline src/snowflake/snowpark/dataframe.py:3503]: consistent comment

File: tests/ast/test_ast_driver.py [hunk 6]
```diff
@@ -146,19 +160,24 @@ def run_test(session, test_source):
...
 """
     # We don't care about the results, and also want to test some APIs that can't be mocked. This suppresses an error
     # that would otherwise be thrown.
     session._conn._suppress_not_implemented_error = True
 
     locals = {"session": session}
     exec(source, locals)
-    base64 = locals["result"]
-    return render(base64), base64
+    base64_batches = locals["result"]
```
Comment (sfc-gh-azwiegincew) [inline tests/ast/test_ast_driver.py:179]: base64_batch – it's a single batch

**Review by sfc-gh-azwiegincew — APPROVED** [review]

File: tests/ast/test_ast_driver.py (hunk 6 above)
Comment (sfc-gh-lspiegelberg) [reply tests/ast/test_ast_driver.py:179]: renamed.

**Pull request merged**